packages/central-server/src/database/MeditrakSyncQueue.js

```diff
@@ -0,0 +1,158 @@
+    this.changeTranslators.survey = change => [change];
+  }
+
+  async createPermissionsBasedView() {
+    await createPermissionsBasedMeditrakSyncQueue(this.models.database);
+    winston.info(`Created permissions_based_meditrak_sync_queue`);
+  }
+
+  async refreshPermissionsBasedView() {
+    try {
+      const start = Date.now();
+      await this.models.database.executeSql(
+        `REFRESH MATERIALIZED VIEW CONCURRENTLY permissions_based_meditrak_sync_queue;`,
+      );
+      const end = Date.now();
+      winston.info(`permissions_based_meditrak_sync_queue refresh took: ${end - start}ms`);
+    } catch (error) {
+      winston.error(`permissions_based_meditrak_sync_queue refresh failed: ${error.message}`);
+    }
+  }
+
+  /**
+   * We need to check for permissions changes when processing survey changes
+   * If the permissions have changed, we need to update all related records
+   */
+  async processSurveyChange(surveyChangeData) {
+    const { new_record: newRecord, old_record: oldRecord, ...surveyChange } = surveyChangeData;
+    if (!oldRecord) {
+      // New survey, so permissions cannot have changed
+      return this.addToSyncQueue(surveyChange);
+    }
+
+    if (!newRecord) {
+      // survey is deleted, no need to bother about permissions associated with it
+      return this.addToSyncQueue(surveyChange);
+    }
+
+    if (
+      newRecord.permission_group_id === oldRecord.permission_group_id &&
+      arraysAreSame(newRecord.country_ids, oldRecord.country_ids)
+    ) {
+      // Permissions haven't changed, just queue the survey change
+      return this.addToSyncQueue(surveyChange);
+    }
+
+    // Permissions have changed, enqueue changes for all related records
+    // to ensure devices sync down those records if they now have permissions
+    const survey = await this.models.survey.findById(surveyChange.record_id);
+    const surveyScreenChanges = (await survey.surveyScreens()).map(record => ({
+      record_type: 'survey_screen',
+      record_id: record.id,
+      type: 'update',
+    }));
+    const surveyScreenComponentChanges = (await survey.surveyScreenComponents()).map(record => ({
+      record_type: 'survey_screen_component',
+      record_id: record.id,
+      type: 'update',
+    }));
+    const questionChanges = (await survey.questions()).map(record => ({
+      record_type: 'question',
+      record_id: record.id,
+      type: 'update',
+    }));
+    const optionSetChanges = (await survey.optionSets()).map(record => ({
+      record_type: 'option_set',
+      record_id: record.id,
+      type: 'update',
+    }));
+    const optionChanges = (await survey.options()).map(record => ({
+      record_type: 'option',
+      record_id: record.id,
+      type: 'update',
+    }));
+
+    const allChanges = [
+      surveyChange,
+      ...surveyScreenChanges,
+      ...surveyScreenComponentChanges,
+      ...questionChanges,
+      ...optionSetChanges,
+      ...optionChanges,
+    ];
```

Comment:
It would be good to run some stress testing on this - does it perform ok if e.g. all of the ssc's are being changed anyway (because they're deleted and created again) so they're also being added to the sync queue through the normal process? We don't hit too much locking if you do that with a bunch of surveys in a row?

Reply:
Good call! I had sort of assumed that this change handler would run after all the associated changes had resolved due to the debounce time, but it's definitely possible that it could kick off halfway which might make things awkward. I'll make a note in the test plan 👍